group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -428,9 +430,44 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        List<CompletableFuture<ListGroupsResponseData>> futures = new java.util.ArrayList<>(Collections.emptyList());
+        for (TopicPartition tp : runtime.partitions()) {
+            futures.add(runtime.scheduleReadOperation(
+                "list_groups",
+                tp,
+                (coordinator, lastCommittedOffset) -> coordinator.listGroups(context, request, lastCommittedOffset)
+            ).exceptionally(exception -> {
+                if (!(exception instanceof KafkaException)) {
+                    log.error("ListGroups request {} hit an unexpected exception: {}",
+                        request, exception.getMessage());
+                }
+                return new ListGroupsResponseData()
+                    .setErrorCode(Errors.forException(exception).code());
+            }));
+        }
+        CompletableFuture<ListGroupsResponseData> responseFuture = new CompletableFuture<>();
+        List<ListGroupsResponseData.ListedGroup> listedGroups = new ArrayList<>();
+        futures.forEach(CompletableFuture::join);
+        for (CompletableFuture<ListGroupsResponseData> future : futures) {
+            try {
+                ListGroupsResponseData data = future.get();
+                if (data.errorCode() != Errors.NONE.code()) {
+                    responseFuture.complete(data);
+                    return responseFuture;
+                }
+                listedGroups.addAll(future.get().groups());
+            } catch (InterruptedException | ExecutionException e) {
+                log.error("ListGroups request {} hit an unexpected exception: {}",
+                    request, e.getMessage());
+                if (!responseFuture.isDone()) {
+                    responseFuture.complete(new ListGroupsResponseData()
+                        .setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code()));
+                    return responseFuture;
+                }
+            }
+        }
```

Comment:
Have you thought about writing an helper method to turn a list of CompletableFutures to a CompletableFuture containing the list of the result? That would be a nice building block that we could put in `FutureUtils`. If any of the CompletableFutures would fail, the resulting CompletableFuture would be failed as well with the same error. This would simplify the code here?

Reply:
I tried to simplify and improve a version based on my own understanding, but I’m not sure if it’s consistent with what you expected.

Reply:
I was thinking about something like this:

```
    public static <T> CompletableFuture<List<T>> allOf(List<CompletableFuture<T>> futures) {
        return CompletableFuture
            .allOf(futures.toArray(new CompletableFuture[0]))
            .thenApply(__ -> {
                final List<T> results = new ArrayList<>(futures.size());
                for (CompletableFuture<T> future : futures) {
                    results.add(future.join());
                }
                return results;
            });
    }
```

I am actually not sure that it would help here, see my previous comment.